pkg/provisioner/ironic/ironic.go

```diff
@@ -1129,6 +1129,91 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	if p.bmcAccess.RAIDInterface() != "" {
+		// Set raid configuration
+		err = setTargetRAIDCfg(p, ironicNode)
+		if err != nil {
+			p.log.Error(err, "failed in setting the target RAID configuration")
+			return
+		}
+		// Build raid clean steps
+		cleanSteps = append(cleanSteps, buildRAIDCleanSteps(p.host.Status.Provisioning.RAID)...)
+	} else if p.host.Status.Provisioning.RAID != nil {
+		return result, fmt.Errorf("the BMC(%v) of the current node does not support RAID", p.bmcAccess.Driver())
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		if !success {
+			err = fmt.Errorf("start manual clean failed")
+		}
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Available, nodes.Active:
+		// Do Nothing (go to provisioning state)
+
+	case nodes.Manageable:
+		if p.host.Status.Cleaned {
+			// Manual clean finished
+			return
+		}
+		result, err = p.startManualCleaning(ironicNode)
+		if err == nil {
+			p.host.Status.Cleaned = true
```

Comment:
That's slightly confusing for me. The name of the flag suggests that the host has been cleaned, rather than a manual action has been executed, as per the comment. Moreover, I think that dirty flag on result must be set in order to record the change. What about instead using Ironic for that?

Reply:
Thanks, I have update the name of **Cleaned** flag to **ManualCleaning**

Reply:
An imminent change (#761) will prohibit the provisioner from writing to the Host. We need to communicate any information like this back to the controller through extra return values.